[MP Rewrite Part 8] Optimize geometry checking

kylestach opened the pull request:
## Description
This optimizes the field geometry checking to use sane rectangle collision algorithms.

## Associated Issue
Issue #1533

kylestach committed:
Optimize rectangle intersection fucntion

kylestach committed:
Make gameplay geometry proper Rects rather than polygons

oswinso commented on common/Geometry2d/Rect.cpp:
```diff
@@ -237,20 +237,17 @@ bool Rect::nearPoint(Point other, float threshold) const {
-        }
-    }
+    double dx = std::max<double>({0.0, max_x - other.x(), other.x() - min_x});
+    double dy = std::max<double>({0.0, max_y - other.y(), other.y() - min_y});
 
-    return false;
+    double distance_to_point = std::sqrt(dx * dx + dy * dy);
+
+    return distance_to_point < threshold;
```
I don't understand how this logic works, can you explain it?

kylestach replied:
I'll add comments, but it's pretty simple. Because the box is axis-aligned, it's easy to find the minimum distance in the `x` and `y` directions separately (at most one of `x - x_max (above), 0 (inside), x_min - x (below)` will be positive, where "positive" represents "outside").

oswinso replied:
ok yeah after you fix the typo it makes sense.

kylestach committed:
Oops, all typos

oswinso commented:
I'm a bit concerned about that typo making it through tests. Can you add a test that checks this?

kylestach commented:
That typo didn't make it through tests, lol. I just never bothered to run them.

oswinso reviewed (approved):
Once you fix the tests, lgtm

kylestach commented:
@oswinso the tests are fine. The failures are for unrelated reasons.

kylestach committed:
[MP Rewrite Part 9] Improve test coverage for motion planning primitives (#1536)

* Write Bezier path tests

* Improve automated testing

* Fix some primitives for tests

[MP Rewrite Part 10] Overall improvements to planners (#1537)

* Add angle overrides and fix pivot planner. Small tweaks to settle.

* Add TODOs for curvature limiting/Bezier spline issues

* Make planner tests deterministic

* Clean up and comment logic in settle planner

* Remove empty CreatePathTest.cpp

* Re-add RRT tests in CreatePathTest

* Make settle planner pick best buffer

This fixes the edge case where the planner can't find an path that finds
the ball quickly enough because the only way to catch it doesn't have
enough buffer time (i.e. immediately before it is caught).

[MP Rewrite Part 11] Miscellaneous formatting and cleanup (#1540)

* Formatting

* More formatting

* Tidy fixes

* Straggling formatting

* Formatting

* Rename low_level to primitives

* Comment on cached_pivot_point

[MP Rewrite Part 12] Port the line kick planner (#1541)

* Add in sketchy line kick

* Formatting

* Tidy

[MP Rewrite Part 13] Stop play fixes and holding trajectory (#1542)

* Fix stopped play and hold in place at end of path move

* Formatting

* Dumb NOLINT

* Style again
